mitxgraders/helpers/calc/expressions.py

```diff
@@ -575,90 +472,101 @@ def parse_algebra(self):
         sumdiff.addParseAction(self.group_if_multiple('sum'))
 
         # Close the recursion
-        expr << sumdiff
-
-        # Save the resulting tree
-        self.tree = (expr + stringEnd).parseString(self.math_expr)[0]
+        expression << sumdiff
 
-    def dump_parse_result(self):  # pragma: no cover
-        """Pretty-print an XML version of the parse_result for debug purposes"""
-        print(self.tree.asXML())
+        return expression + stringEnd
 
-    def set_vars_funcs(self, variables=None, functions=None):
-        """Stores the given dictionaries of variables and functions for future use"""
-        self.vars = variables if variables else {}
-        self.functions = functions if functions else {}
-
-    def evaluate(self, allow_inf):
+    def raw_parse(self, expression):
         """
-        Recursively evaluate `self.tree` and return the result.
+        Try to parse a string and cache the result. ALWAYS clears storage.
         """
-        def handle_node(node):
-            """
-            Return the result representing the node, using recursion.
-
-            Call the appropriate action from self.actions for this node. As its inputs,
-            feed it the output of `handle_node` for each child node.
-            """
-            if not isinstance(node, ParseResults):
-                # Entry is either a (python) number or a string.
-                # Return it directly to the next level up.
-                return cast_np_numeric_as_builtin(node)
+        try:
+            BracketValidator.validate(expression)
+            tree = self.grammar.parseString(expression)[0]
+            parsed = MathExpression(expression,
```

Comment:
I really like this. Having separated out the suffixes, tree parsing is indeed general.